### PR TITLE
Changed warning in particle generator to be more informative when

### PR DESCRIPTION
**Pull request by bruttenberg**
sampling during factored algorithms. Also updated particle generator
with more informative variable names, deprecated old names, and updated
related files.

### CODE REVIEW

File: Figaro/src/main/scala/com/cra/figaro/algorithm/factored/factors/factory/ChainFactory.scala [hunk 1]
```diff
@@ -62,5 +71,44 @@ object ChainFactory {
...
+        // Need to normalize subsolution in case there's any nested evidence
+        val subsolution = subproblem.solution.reduce(_.product(_))
+        val sum = subsolution.foldLeft(subsolution.semiring.zero, subsolution.semiring.sum(_, _))
+        val subVars = subsolution.variables
+        if (subVars.length == 1) {
+          val subVar = subVars(0)
+          for { subVal <- subVar.range } {
+            val childIndex = childVar.range.indexOf(subVal)
+            val subIndex = subVar.range.indexOf(subVal)
+            val entry = subsolution.semiring.product(subsolution.get(List(subIndex)), 1.0 / sum)
```
Comment (bruttenberg) [inline Figaro/src/main/scala/com/cra/figaro/algorithm/factored/factors/factory/ChainFactory.scala]: This is not semiring safe. The 1.0 should be replaced by semiring.one, and the divide should be done through a semiring operation. However, this is only valid for divideable semirings. @apfeffer please review this.

**Comment by apfeffer** [comment]
Hmm. This is a normalization operation. Do we need to normalize?

**Comment by bruttenberg** [comment]
We probably don’t need to normalize, but I’m not 100% sure. William noted though that our factor creation code is in general not semiring safe, so this might not matter in the larger context.

+  +  +  +  +  +  +  +  +  +  +  +  +  +
Brian Ruttenberg, PhD
Senior Scientist
Decision Management Systems Division
Government Services
Charles River Analytics Inc.
617.491.3474 x730
www.cra.com<http://www.cra.com/>

**Comment by apfeffer** [comment]
The important question is whether this construction works for MAP queries.

**Comment by bruttenberg** [comment]
I’d have to look more. FYI, this is really the only open issue for 4.1. I made minor changes to your new chain stuff.

+  +  +  +  +  +  +  +  +  +  +  +  +  +
Brian Ruttenberg, PhD
Senior Scientist
Decision Management Systems Division
Government Services
Charles River Analytics Inc.
617.491.3474 x730
www.cra.com<http://www.cra.com/>

**Pull request merged**